tests/subsys/zbus/channel_id/src/main.c

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright (c) 2024 Embeint Inc
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/zbus/zbus.h>
+#include <zephyr/ztest.h>
+#include <zephyr/ztest_assert.h>
+
+struct msg {
+	int x;
+};
+
+enum channel_ids {
+	CHAN_A = 100,
+	CHAN_B = 123,
+	CHAN_C = 0x12343243,
+	CHAN_D = 123,
+	CHAN_E = 1,
+	CHAN_F = 357489,
+};
+
+ZBUS_CHAN_ID_DEFINE(chan_a, CHAN_A, struct msg, NULL, NULL, ZBUS_OBSERVERS_EMPTY, ZBUS_MSG_INIT(0));
+ZBUS_CHAN_ID_DEFINE(chan_b, CHAN_B, struct msg, NULL, NULL, ZBUS_OBSERVERS_EMPTY, ZBUS_MSG_INIT(0));
+ZBUS_CHAN_ID_DEFINE(chan_c, CHAN_C, struct msg, NULL, NULL, ZBUS_OBSERVERS_EMPTY, ZBUS_MSG_INIT(0));
+ZBUS_CHAN_ID_DEFINE(chan_d, CHAN_D, struct msg, NULL, NULL, ZBUS_OBSERVERS_EMPTY, ZBUS_MSG_INIT(0));
+ZBUS_CHAN_ID_DEFINE(chan_e, CHAN_E, struct msg, NULL, NULL, ZBUS_OBSERVERS_EMPTY, ZBUS_MSG_INIT(0));
+ZBUS_CHAN_ID_DEFINE(chan_f, CHAN_F, struct msg, NULL, NULL, ZBUS_OBSERVERS_EMPTY, ZBUS_MSG_INIT(0));
+ZBUS_CHAN_DEFINE(chan_g, struct msg, NULL, NULL, ZBUS_OBSERVERS_EMPTY, ZBUS_MSG_INIT(0));
+
+ZTEST(channel_id, test_channel_retrieval)
+{
+	/* Invalid/unknown channel IDs */
+	zassert_is_null(zbus_chan_from_id(0x1000000));
+	zassert_is_null(zbus_chan_from_id(ZBUS_CHAN_ID_INVALID));
+
+	/* Standard retrieval */
+	zassert_equal(&chan_a, zbus_chan_from_id(CHAN_A));
+	zassert_equal(&chan_c, zbus_chan_from_id(CHAN_C));
+	zassert_equal(&chan_e, zbus_chan_from_id(CHAN_E));
+	zassert_equal(&chan_f, zbus_chan_from_id(CHAN_F));
```

Comment:
It is not desired flexibility, but a consequence of the fact we have no way to ensure at compile-time there are no duplicate IDs.

It is therefore possible for the IDs to be duplicated, so some sort of behavior needs to be specified.
I have chosen "either channel with the ID will be returned".

Reply:
I agree, but I imagine it would be adequate to raise an assertion error when it happens. It should not be considered normal.

Reply:
The challenge with assertions (in my opinion) is that in many applications the assertions aren't compiled in, and are in fact discouraged by the build system outside of tests.
https://github.com/zephyrproject-rtos/zephyr/blob/65f63622e273c19883b866d71a13e8f18b7d5c6c/CMakeLists.txt#L2093-L2097

Specifying that the behavior is "raises an assertion" for me means that most of the time the behavior is "not handled".
In the `init` function I have added the assert case, with an error message stating the duplicated ID that (only compiled in when asserts are disabled).

What behavior do you want to see in this test, given that B and G have the same ID (to validate the error is printed)?

Reply:
I like this [text](https://interrupt.memfault.com/blog/defensive-and-offensive-programming) to explain my fondness for assertions. I wouldn't say I like to start developing using defensive programming. The assertion will "guarantee" that developers would not have simple defects in their code. I don't see the message about asserts as "only for tests." It is for development and not production. I rely a lot on assertions on zbus to give developers good and quick feedback about something they are doing incorrectly or inappropriately. A duplicated ID is a case the developer must be aware of as soon as possible to fix, and don't waste time trying to solve some weird behavior because of that. 

> What behavior do you want to see in this test, given that B and G have the same ID (to validate the error is printed)?

I am not sure you can test the assertion results in the init function. For the regular zbus functions, I have created the `ZBUS_ASSERT` that returns `-EFAULT` during tests when the `ZBUS_ASSERT_MOCK` is enabled. In that case, you could return it when a duplication was detected. But inside the init function, I don't know if there is a way.

Reply:
The assert exists in the `init` function, and if assertions are disabled there is the error log. Both perform the task you desire, quick feedback for development.
I'm not arguing assertions == bad, simply that given the test contains duplicated IDs, we have to choose "something".

* Test `zbus_chan_from_id` with the duplicated ID 
* Don't test `zbus_chan_from_id` with the duplicated ID

If we are testing it with the duplicated ID, we need to choose what it should be returning.

* `NULL` (requires iterating over the entire channel list in every call to detect duplicates)
* First matching channel (no overhead in the "normal" case, minimally wrong if duplicate ID's do exist)

I chose to test it and return the first matching as it is lower overhead for the happy path.
If you think a different choice should be made for the test, I am happy to hear your reasoning.

This is an orthogonal discussion to the assert however, as that is already implemented.

Reply:
I see. We need to test that. You are doing that ok. The only attention point is that you are not covering the assert code portion with tests. Are you planning to change some samples to add the ID feature? If so, you could have a version of that with duplicated IDs and try to capture the ASSERTION error string with the `harness: console`. What do you think?

Reply:
@JordanYates?

Reply:
What is the "assert code portion"? This PR does not introduce any new usage of `__ASSERT`.